npep/npep-137-conformance-profiles.md

```diff
@@ -0,0 +1,187 @@
+`StandardExtendedFeatures`. The features that fall under the `StandardCoreFeatures`
+are a must for the implementations to be conformant while the features that fall
+under `StandardExtendedFeatures` are opt-ins which are not mandatory.
+
+In this NPEP, we will add a concept called `named profiles` which will indicate a
+"level of conformance" on a resource level. We are picking profiles on a resource
+level because that is what makes sense for the project as of today. This can be
+changed as the project evolves. Using these `named profiles`, implementations can
+run the conformance tests and prove they satisfy the requirements for each profile
+and receive conformance badges from the Network Policy API project.
+
+This NPEP aims to provide an API (CRD) for reporting results from these conformance
+tests. CLI tooling will also be provided for invoking the conformance tests with
+the profile names so that the results can be easily submitted back upstream.
+
+## User-Stories/Use-Cases
+
+Story 1: Receive an API conformance badge
+
+As a CNI maintainer, I want to receive recognition from Network Policy API repo
+that my plugin implements `AdminNetworkPolicy` API. This badge will enable us
+to advertise upstream feature support in our plugin easily to our users.
+
+Story 2: Track Implementations for each API resource
+
+As a Network Policy project maintainer, I would like to easily track which
+implementations are using the defined API resources in our project so that it
+paves way for easier collaboration and feedback loops when changes are introduced
+to the API in each release.
+
+## API
+
+We will add a new API resource called `ConformanceReport` which will be at
+the center of our test result reporting workflow. The implementors running
+the tests can then:
+
+1. Choose a `named profile`
+2. Add integration for the tests under this profile in their downstream project
+3. Report the results to Network Policy API project using this new API resource
+4. Get a conformance badge recognition via our official website
+
+## Profiles
+
+Named Profiles for Network Policy API project will be tied to each API resource.
+We will start with two named profiles and expand this as the project evolves.
+
+1. AdminNetworkPolicy
+2. BaselineAdminNetworkPolicy
+
+Each of these profiles may have a combination of conformance tests that fall under
+`StandardCoreFeatures` and `StandardExtendedFeatures`. Example; if I pick the profile
+`AdminNetworkPolicy`, all tests like `AdminNetworkPolicyEgressSCTP` and
+`AdminNetworkPolicyPriorityField` fall under the `SupportAdminNetworkPolicy` feature
+which is under the `StandardCoreFeatures` subset. So these tests must pass for the
+`AdminNetworkPolicy` profile conformance. Whereas tests (we don't have any today)
+that fall under `AdminNetworkPolicySameLabels` feature which is under the
+`StandardExtendedFeatures` subset are not mandatory for `AdminNetworkPolicy`
+profile conformance.
```

Comment:
Is there a way to report all of the following:

"I support ONLY the standard features"
"I support All the standard features and some extended"
"I support ALL features"

Reply:
good question, so the report generates something like this:
```
profiles:
        - core:
            failedTests:
            - AdminNetworkPolicyIngressSCTP
            - AdminNetworkPolicyEgressUDP
            result: failure
            statistics:
              Failed: 2
              Passed: 5
              Skipped: 0
            summary: ""
          name: AdminNetworkPolicy
        - core:
            result: success
            statistics:
              Failed: 0
              Passed: 7
              Skipped: 0
            summary: ""
          name: BaselineAdminNetworkPolicy
       - extended:
           result: success
           statistics:
             Failed: 0
             Passed: 2
            Skipped: 0
           name: AdminNetworkPolicySameLabels
```
from here we can infer does this support ALL features (all pass), ONLY Core (only those will be reported as pass), mix of Core and Extended, and which ones are failing...

Reply:
Sweet! That's what I was looking for :)